Initial integration with shelley node

Anviking opened the pull request:
# Issue Number

ADP-289, #1657

# Overview

- [x] Copied the `cardano-wallet-byron` structure and made it compile for shelley
- [x] Completely unimplemented `TransactionLayer`
- [x] A few integration tests and unit tests are running (that don't need txs)
- [x] Add appropriate CLI and API
- [x] Make sure a wallet can be restored

# Comments

<!-- Additional comments or screenshots to attach if any -->

<!-- 
Don't forget to:

 ✓ Self-review your changes to make sure nothing unexpected slipped through
 ✓ Assign yourself to the PR
 ✓ Assign one or several reviewer(s)
 ✓ Once created, link this PR to its corresponding ticket
 ✓ Assign the PR to a corresponding milestone
 ✓ Acknowledge any changes required to the Wiki
-->

Anviking committed:
Scaffold cardano-wallet-shelley from the byron one

KtorZ commented on lib/shelley/exe/cardano-wallet-shelley.hs:
```diff
@@ -0,0 +1,309 @@
+        <> cmdKey
+        <> cmdWallet cmdByronWalletCreate byronWalletClient
+        <> cmdAddress byronAddressClient
+        <> cmdTransaction byronTransactionClient byronWalletClient
```
These will have to be change to point to the "normal" API. We could perhaps still have them behind a `legacy-wallet` and `legacy-address` sets of command.

Anviking replied:
> These will have to be change to point to the "normal" API.

Done.

KtorZ commented on lib/shelley/src/Cardano/Wallet/Shelley/Api/Server.hs:
```diff
@@ -0,0 +1,256 @@
+        :<|> (\_ -> throwError err501)
+        :<|> (\_ _ -> throwError err501)
+        :<|> (\_ _ -> throwError err501)
+        :<|> (\_ -> throwError err501)
```
These will now be accessible in Shelley.

Anviking replied:
The shelley wallet endpoints are now added.

KtorZ commented on lib/shelley/src/Cardano/Wallet/Shelley/Api/Server.hs:
```diff
@@ -0,0 +1,256 @@
+    :<|> proxy
+  where
+    byron :: ApiLayer (RndState n) t ByronKey
+    byron = error "todo: should we have a byron api here?"
```
Yes.

KtorZ commented on lib/shelley/src/Cardano/Wallet/Shelley/Api/Server.hs:
```diff
@@ -0,0 +1,256 @@
+        , PaymentAddress n IcarusKey
+        , PaymentAddress n ByronKey
+        )
+    => ApiLayer (SeqState n IcarusKey) t IcarusKey
```
Similarly to Jörmungandr, we'll have to instantiate 3 ApiLayer: Byron, Icarus and Shelley. 
Although, what we currently call `Shelley` really is `Jormungandr` so it'll have to be renamed (and moved) first.

Anviking replied:
Should be done.

KtorZ commented on lib/shelley/src/Cardano/Wallet/Shelley/Compatibility.hs:
```diff
@@ -0,0 +1,454 @@
+toGenTx :: HasCallStack => W.SealedTx -> GenTx (ShelleyBlock c)
+toGenTx = error "todo: toGenTx not implemented"
+
+fromShelleyTx :: SL.Tx c -> W.Tx
```
minor note: the conventions I have been trying to enforce in this file were that:

a) Identifiers from cardano-wallet are used qualified.
b) Identifiers from networking, consensus or ledger were used unqualified.  
c) Function are named after the data-type
d) `toXXX` when they translate from a cardano-wallet type to a networking, consensus & ledger data-type. `fromXXX` when they do the opposite (from networking to cardano-wallet).

Anviking replied:
ouroboros-network seems to consistently imports everything from the Shelley ledger as qualified `as SL`.

There are name clashes (e.g. `Block`, `Tx`) between ouroboros-network and the specs.

I think we have to use `qualified` to some extent. At least for `SL` and `W`.

KtorZ commented on lib/shelley/src/Cardano/Wallet/Shelley/Compatibility.hs:
```diff
@@ -0,0 +1,454 @@
+                SL.bheaderPrev header
+            }
+        , transactions = txs
+        , delegations  = []
```
Is this always the case for now or are delegation certificates somewhat already specified in the ShelleyBlock type?

Anviking replied:
Looks like we can access them

https://github.com/input-output-hk/cardano-ledger-specs/blob/caa0b377698aace2550d1a9948a050bfa296a3e0/shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/TxData.hs#L315-L328

KtorZ commented on lib/shelley/src/Cardano/Wallet/Shelley/Network.hs:
```diff
@@ -0,0 +1,962 @@
+-- - Module's documentation in `ouroboros-network/typed-protocols/src/Network/TypedProtocols.hs`
+-- - Data Diffusion and Peer Networking in Shelley (see: https://raw.githubusercontent.com/wiki/input-output-hk/cardano-wallet/data_diffusion_and_peer_networking_in_shelley.pdf)
+--     - In particular sections 4.1, 4.2, 4.6 and 4.8
+module Cardano.Wallet.Shelley.Network
```
Make sure to base this "fork" on the latest version of the Byron Network code which includes Rodney's latest refactoring and inclusion of the local state query (you can use a "null client" here if the protocol isn't available yet on the node's end, but at least, we have the skeleton in place on the wallet's side).

Anviking replied:
> if the protocol isn't available yet on the node's end

It was, was I was grepping wrong.

KtorZ commented on lib/shelley/src/Cardano/Wallet/Shelley/Transaction.hs:
```diff
@@ -0,0 +1,116 @@
+    -- TODO: Not sure why we need this function, but should we call blake2b256
+    -- for Shelley really?
+    mkTx out@(TxOut (Address bytes) _) =
+        Tx (Hash $ blake2b256 bytes) [] [out]
```
This function is needed for our integration test setup. With Jörmungandr, the genesis file is mapped to a genesis block which contains a set of initial transaction (although with no inputs). On cardano-node however, the genesis contains an initial set of Genesis TxOut, so this function artificially maps that to an initial "Block". 

May you clarify this with a comment for future readers?

KtorZ commented on lib/shelley/src/Cardano/Wallet/Shelley/Transaction/Size.hs:
```diff
@@ -0,0 +1,234 @@
+import qualified Codec.CBOR.Write as CBOR
+import qualified Data.ByteString as BS
+import qualified Data.ByteString.Lazy as BL
+
```
I assume none of these :point_down: will apply to Shelley on-chain data-types. We should adjust that based on the CDDL spec.

Anviking replied:
Can do when adding tx support, I'd say.

KtorZ replied:
Yes, but in the meantime, I'd not add this at all since we know it doesn't apply already.

KtorZ commented on lib/shelley/test/data/cardano-node/node.cert:
```diff
@@ -0,0 +1,8 @@
+    "b8j8qcnO/HyB5PcDgROOrF3fGHyAe1Ec2vitqVas+ZfWdrcaTfRDIhcKT+6jsNh0oQuKi7I23shW+1X7V3Y2Pw=="
+, "cert":
+    "ea983bfd989be8d1ac1529e0d9f9ef723766397ad58ea24b89d09c40fc3122a8728f3efaca35f16ab994aa02fff76098006889f20810c40d47a34a3a6b9d3e05"
+}
```
I'd assume that this only applies to Byron. Shelley uses VRF/KES credentials for stake pools.

KtorZ commented on lib/shelley/test/integration/Cardano/Wallet/Shelley/Faucet.hs:
```diff
@@ -0,0 +1,21 @@
+{-# LANGUAGE DataKinds #-}
+
+module Cardano.Wallet.Byron.Faucet
```
*Shelley.Faucet

Anviking committed:
Prep: Rename fromGenesisTxOut -> genesisBlockFromTxOuts and document

disassembler committed:
cardano-wallet-shelley: initial nix configuration

* adds to default.nix: shell, top-level attr and dockerImage

KtorZ commented on lib/byron/src/Cardano/Wallet/Byron/Transaction.hs:
```diff
@@ -16,7 +16,7 @@
 module Cardano.Wallet.Byron.Transaction
     ( newTransactionLayer
-    , fromGenesisTxOut
+    , genesisBlockFromTxOuts
```
:+1:

KtorZ commented on lib/shelley/src/Cardano/Wallet/Shelley.hs:
```diff
@@ -0,0 +1,464 @@
+-- the wallet.
+serveWallet
+    :: forall t c.
+        ( t ~ IO (Shelley c)
```
Is there any value for us to carry this `c` parameter around? Certainly, we can already make a choice for the underlying Crypto scheme (for instance: `TPraosStandardCrypto`) as it seems to be done in:

https://github.com/input-output-hk/cardano-node/blob/master/cardano-api/src/Cardano/Api/Types.hs#L127-L146

Anviking replied:
Was wondering the same. Went with it to see how it would work out. Maybe we would want to use some mock crypto for… performance at some point?

But certainly we can remove it.

KtorZ replied:
When the crypto becomes our bottleneck in performances here, we would have come a long way :grimacing:

KtorZ commented on lib/shelley/src/Cardano/Wallet/Shelley/Compatibility.hs:
```diff
@@ -0,0 +1,443 @@
+import Cardano.Chain.Common
+    ( unsafeGetLovelace )
+import Cardano.Chain.UTxO
+    ( Tx (..), TxAux, TxIn (..), TxOut (..), taTx )
```
I don't think any of these apply to `Shelley`. 
We should try to use things from https://github.com/input-output-hk/cardano-node/blob/master/cardano-api/src/Cardano/Api/Types.hs#L108-L114, or better, use `Cardano.Api` here to interface with cardano-wallet

Anviking replied:
👍 I simply removed them for now.

KtorZ commented on lib/shelley/src/Cardano/Wallet/Shelley/GenesisDataOrphans.hs:
```diff
@@ -0,0 +1,197 @@
+{-# OPTIONS_GHC -Wno-missing-methods #-}
+{- HLINT ignore -}
+
+-- Module copied from cardano-node.
```
Couldn't the module be imported instead so that the instances are put in Context?

Anviking replied:
I got the sense that depending on cardano-node would be cumbersome due to all the dependencies.

Anviking replied:
Done, wasn't that cumbersome, but needed to add a new Iohk-monitoring plugin to the snapshot.

KtorZ commented on lib/shelley/src/Cardano/Wallet/Shelley/Network.hs:
```diff
@@ -0,0 +1,1101 @@
+--      *-----------*
+--      | Intersect |◀══════════════════════════════╗
+--      *-----------*         FindIntersect         ║
+--            │                            :750
```
:thinking:

Anviking replied:
👍Fixed and changed the commits to be more
1. Copy byron implementation
2. Tweak

for nicer diffs.

Anviking committed:
Prep: Share Byron Address integration tests in core-integration

Rather than duplicating them in both byron and shelley

Anviking committed:
Prep: Share Scenario.CLI.Byron.Wallets in core-integration

Anviking committed:
Prep: More detailed (Buildable BlockHeader) instance

Now:
5df6aa57<-[37f92da1-0.3933#207]

Previously:
37f92da1-[0.3933#207]

Anviking commented on lib/shelley/src/Cardano/Wallet/Shelley/Network.hs:
```diff
@@ -698,131 +670,98 @@ chainSyncWithBlocks
-        --     |---------------------------|-------------->
-        --               pipelined            one by one
-        --
-        pipeline
```
When I started, pipelining wasn't on master. 

Should be simple to revert this deletion now.

Could also factor out the logic to a block-agnostic shared location… or maybe in the next PR.

Anviking commented on lib/shelley/src/Cardano/Wallet/Shelley/Network.hs:
```diff
@@ -1132,6 +1070,12 @@ instance HasSeverityAnnotation NetworkLayerLog where
         MsgPostSealedTx{}          -> Debug
         MsgLocalStateQuery{}       -> Debug
-        MsgNodeTip{}               -> Debug
+        MsgNodeTip{}               -> Info
```
Should revert this and similar things if any.

Anviking committed:
Add Shelley Implementation

- Replace duplicated Byron implementation with compiling and working one
- Integration test setup
- Add Shelley API
- Make CLI use Shelley specific commands instead of Byron

fixup: remove start.sh since it is not currently wokring

fixup use default codec in shelley

fixup weeder

Add cardano-node to nix env for shelley tests

fixup hlint

Regenerate nix

Fixup: Remove shelley TransactionSpec

Fixup integration genesis to produce blocks

Anviking committed:
Prep: Use deserialised tips in follow-tip-chain-sync

This allows us to use the default codecs instead of custom ones.

Anviking committed:
Use default codecs in Byron Network module

KtorZ commented on lib/byron/src/Cardano/Wallet/Byron/Network.hs:
```diff
@@ -378,6 +354,12 @@ mkWalletClient bp chainSyncQ =
+    codecs :: MonadST m => ClientCodecs ByronBlock m
+    codecs = clientCodecs
+        (ByronCodecConfig (toEpochSlots getEpochLength))
+        ByronNodeToClientVersion1
```
Aren't we using `Version2` ?

Anviking replied:
Hm… I was going to say that this is another kind of version… but yes — we should probably use https://github.com/input-output-hk/ouroboros-network/blob/02a808741ef3fc8b244c4be221616e4fc92a8310/ouroboros-consensus-byron/src/Ouroboros/Consensus/Byron/Ledger/NetworkProtocolVersion.hs#L32

Anviking replied:
Done

KtorZ commented on lib/byron/src/Cardano/Wallet/Byron/Network.hs:
```diff
@@ -458,20 +427,24 @@ mkTipSyncClient tr gbp localTxSubmissionQ onTipUpdate onTxParamsUpdate = do
-                    nodeDecodeQuery
-                    nodeEncodeResult
-                    nodeDecodeResult
+                codec = cStateQueryCodec codecs
```
Sweet. Nice simplification.

KtorZ commented on lib/shelley/src/Cardano/Wallet/Shelley.hs:
```diff
@@ -0,0 +1,454 @@
+            ( NetworkDiscriminantVal n
+            , PaymentAddress n IcarusKey
+            , PaymentAddress n ByronKey
+            , PaymentAddress n ShelleyKey
```
This is incorrect at the moment, but we can solve later. We need to:

a) Rename 'ShelleyKey' to 'JormungandrKey'
b) Port https://github.com/input-output-hk/cardano-addresses/blob/master/src/Cardano/Address/Style/Shelley.hs

KtorZ commented on lib/shelley/src/Cardano/Wallet/Shelley/Compatibility.hs:
```diff
@@ -0,0 +1,413 @@
+    { getFeePolicy = W.LinearFee
+        (Quantity (naturalToDouble $ SL._minfeeA pp))
+        (Quantity (fromIntegral $ SL._minfeeB pp))
+        (Quantity 0) -- TODO: it's not as simple as this?
```
No it's not indeed. To be addressed when we re-introduce certificates.

KtorZ commented on lib/shelley/src/Cardano/Wallet/Shelley/Compatibility.hs:
```diff
@@ -0,0 +1,413 @@
+    W.ProtocolMagic (fromIntegral magic)
+
+fromProtocolMagicId :: ProtocolMagicId -> W.ProtocolMagic
+fromProtocolMagicId = W.ProtocolMagic . fromIntegral . unProtocolMagicId
```
I thought the concept of protocolMagic was gone in Shelley :thinking: ? This one is probably unused.

Anviking replied:
Done

KtorZ commented on lib/shelley/src/Cardano/Wallet/Shelley/Launch.hs:
```diff
@@ -0,0 +1,373 @@
+data NetworkConfiguration where
+    MainnetConfig
+        :: (SomeNetworkDiscriminant, NodeVersionData)
+        -> NetworkConfiguration
```
I wonder if we shouldn't remove anything regarding "Shelley mainnet" at this stage.

Anviking replied:
Good point, probably. Done.

KtorZ commented on lib/shelley/test/integration/Main.hs:
```diff
@@ -0,0 +1,220 @@
+            describe "Miscellaneous CLI tests" $ parallel (MiscellaneousCLI.spec @t)
+            describe "Key CLI tests" $ parallel (KeyCLI.spec @t)
+        describe "API Specifications" $ specWithServer tr $ do
+            Network.spec
```
What about the wallet specs?

Anviking replied:
They need transactions. For faucet txs I think.

KtorZ replied:
> For faucet txs I think.

Precisely not :) The faucets are genesis UTxOs, so most of the wallet scenarios are actually about restoring wallets from on-chain data.

Anviking replied:
Ok. The node has stopped producing blocks for some reason 🤔

Anviking replied:
There are a few scenarios requiring transactions. I can't mark them pending since they exist in core-integration. Let's deal with this later.

KtorZ commented on lib/shelley/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs:
```diff
@@ -0,0 +1,1124 @@
+{-# LANGUAGE TypeApplications #-}
+
+{-# OPTIONS_GHC -fno-warn-orphans #-}
+module Cardano.Wallet.Shelley.TransactionSpec
```
Most of this module is irrelevant for Shelley since the transaction format has completely changed since Byron. I'd suggest to remove it; it'll be re-implement once we implement the transaction layer for Shelley.

Anviking replied:
48cf00bd5

KtorZ commented on stack.yaml:
```diff
@@ -1,4 +1,4 @@
-resolver: https://raw.githubusercontent.com/input-output-hk/cardano-haskell/master/snapshots/cardano-1.11.0.yaml
+resolver: https://gist.githubusercontent.com/Anviking/3c48b7343002d58f218b87436bfdaf0c/raw/877a252c197b36632dc52f14370fae3f3ea9c834/cardano-1.12.2.yaml
```
Let's have this as a pre-1.12.0 snapshot on cardano-haskell

Anviking replied:
Done.

Anviking committed:
Adjust stack snapshot

Anviking committed:
Fixup: remove fromProtocolMagicId

Anviking committed:
Fixup: remove misleading shelley MainnetConfig

Anviking committed:
Fixup: Use ByronNodeToClientVersion2

Commit:
Regenerate nix

rvl committed:
docker: tag and push image for shelley

rvl committed:
nix: Add release builds for shelley wallet

rvl committed:
nix: Add shelley to windows testing bundle

rvl committed:
nix: cardano-node 1.11.0 -> latest master (pre 1.12.0 6de6e47)

rvl committed:
Update version to 2020.5.6 in cardano-wallet-shelley.cabal

Commit:
fixup! Regenerate nix

rvl committed:
fixup! nix: Add release builds for shelley wallet

Anviking committed:
Implement fromShelleyTx and toGenTx
